projects/bootloader/src/bootloader_crc32.c

```diff
@@ -0,0 +1,34 @@
+#include "bootloader_crc32.h"
+
+#include "stdio.h"
+
+#include "bootloader_mcu.h"
+#include "crc32.h"
+#include "flash.h"
+
+// calculates the crc32 for the full address, with 2048 byte increments
+// at a time, returning the final crc32 code
+uint32_t calculated_application_crc32(uintptr_t address, size_t size) {
```

Comment:
I've just looked over the implementation of `flash_read`, and it looks like it just `memcpy`s data to the buffer, reading from the flash address it's passed directly. In that case, I think you might be able to just do
```c
crc32_arr(BOOTLOADER_APPLICATION_START, BOOTLOADER_APPLICATION_SIZE)
```
to compute the crc32 of the whole application (which sort of makes this function obsolete, sorry about that).

Test that and see if it works, and then if it does, you might want to just implement this function as the `crc32_arr` call above (and just don't take the `address` and `size` parameters, making this function just a shortcut to compute the crc32 of the application).